File: server/domain/paddle.py
from dataclasses import dataclass
import numpy as np
from domain.ball import Ball


@dataclass
class Paddle:
    INITIAL_Y = 0.5  # Default center position
    
    x_position: float # Position as percentage of screen height (0-1)
    y_position: float = INITIAL_Y  # Position as percentage of screen height (0-1)
    height: float = 0.2  # Height as percentage of screen height
    width: float = 0.02 # Width of the paddle
    speed: float = 0.01  # Movement speed per frame

    def __init__(self, x_pos: float):
        self.x_position = x_pos
        self.h = self.height / 2

    @property
    def y_min(self) -> float:
        return self.y_position - self.h

    @property
    def y_max(self) -> float:
        return self.y_position + self.h

    def is_on_paddle(self, ball: Ball) -> bool:
        if np.abs(ball.x - self.x_position) <= ball.radius + self.width / 2:
            if self.y_min - ball.radius <= ball.y <= self.y_max + ball.radius:
                return True
        return False

    def move_up(self) -> None:
        new_y = self.y_position - self.speed
        self.y_position = max(self.h, new_y)

    def move_down(self) -> None:
        new_y = self.y_position + self.speed 
        self.y_position = min(1.0 - self.h, new_y)

    def reset_position(self) -> None:
        """Reset paddle to center position"""
        self.y_position = self.INITIAL_Y

File: server/game/paddle.go
package game

import "math"

type Paddle struct {
	X      float64
	Y      float64
	Height float64
	Width  float64
	Speed  float64
}

func NewPaddle(xPos float64) *Paddle {
	return &Paddle{
		X:      xPos,
		Y:      0.5,
		Height: 0.2,
		Width:  0.02,
		Speed:  0.01,
	}
}

func (p *Paddle) YMin() float64 {
	return p.Y - p.Height/2
}

func (p *Paddle) YMax() float64 {
	return p.Y + p.Height/2
}

func (p *Paddle) IsOnPaddle(ball *Ball) bool {
	if math.Abs(ball.X-p.X) <= ball.Radius+p.Width/2 {
		if p.YMin()-ball.Radius <= ball.Y && ball.Y <= p.YMax()+ball.Radius {
			return true
		}
	}
	return false
}

func (p *Paddle) MoveUp() {
	newY := p.Y - p.Speed
	p.Y = math.Max(p.Height/2, newY)
}

func (p *Paddle) MoveDown() {
	newY := p.Y + p.Speed
	p.Y = math.Min(1.0-p.Height/2, newY)
}

func (p *Paddle) ResetPosition() {
	p.Y = 0.5
}
